program/src/token.rs

```diff
@@ -22,12 +22,19 @@ use std::{
     ops::{Add, Div, Mul, Sub},
 };
 
-#[derive(Copy, Clone)]
+#[derive(Copy, Clone, PartialEq, Debug)]
 pub struct Rational {
     pub numerator: u64,
     pub denominator: u64,
 }
 
+impl PartialOrd for Rational {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        (self.numerator as f64 / self.denominator as f64)
+            .partial_cmp(&(other.numerator as f64 / other.denominator as f64))
```

Comment:
This comparison is lossy, `f64` cannot represent every `u64` exactly. This can lead to incorrect results:

```rust
#[test]
fn test_rational_cmp() {
    // Construct x and y such that x < y.
    let x = Rational {
        numerator: 1 << 53,
        denominator: 1,
    };
    let y = Rational {
        numerator: x.numerator + 1,
        denominator: x.denominator,
    };
    assert_eq!(x.partial_cmp(&y), Some(std::cmp::Ordering::Less));
}
```
```
thread 'token::test_rational_cmp' panicked at 'assertion failed: `(left == right)`
  left: `Some(Equal)`,
 right: `Some(Less)`', program/src/token.rs:49:5
```

We can fix this by avoiding floats. Observe that
```
   a/b < x/y
=> a   < bx/y    if b > 0
=> ay  < bx      if y > 0
```
We only work with non-negative denominators so the comparison flipping is not an issue, we only have to check to avoid division by zero:

```rust
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        if self.denominator == 0 || other.denominator == 0 {
            None
        } else {
            let x = self.numerator as u128 * other.denominator as u128;
            let y = other.numerator as u128 * self.denominator as u128;
            Some(x.cmp(&y))
        }
    }
```

The multiplications don't overflow because we widen to `u128`.

Please also add tests for other cases (like division by zero).

Reply:
Thanks for the tests and more correct implementation, I'll change it and do more tests 👍